### PR TITLE
Feat/dynamic menu

### PR DESCRIPTION
**Pull request by thomaslennan**
## Description
N-level menu defined via JSON language

Need feedback on the proposed functionality and model definition. Need some styling tips.

### What's included?
New td-dynamic-menu element
Demo
Test
Readme

#### Test Steps
- [ ] `npm run serve`
- [ ] Open demos, navigate to Components > Navigation > Dynamic Menu
- [ ] Confirm expected functionality in demo

#### General Tests for Every PR

- [ ] `npm run serve:prod` still works.
- [ ] `npm run tslint` passes.
- [ ] `npm run stylelint` passes.
- [ ] `npm test` passes and code coverage is not lower.
- [ ] `npm run build:lib` still works.

##### Screenshots or link to StackBlitz/Plunker

Screen Shot 2020-08-10 at 4 47 34 PM
Screen Shot 2020-08-10 at 4 48 02 PM
Screen Shot 2020-08-10 at 4 48 13 PM

### CODE REVIEW

File: src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      <td-dynamic-menu-item #menu [items]="item.children"></td-dynamic-menu-item>
+    </ng-container>
+    <ng-container *ngIf="!item.children || item.children.length === 0">
+      <a *ngIf="item.newTab" mat-menu-item id="{{ item.id }}" [href]="item.link" target="_blank">
+        <mat-icon *ngIf="item.svgIcon" [class]="item.iconClasses" [svgIcon]="item.svgIcon"></mat-icon>
+        <mat-icon *ngIf="item.icon" [class]="item.iconClasses">{{ item.icon }}</mat-icon>
+        <span *ngIf="item.text">
+          {{ item.text }}
+        </span>
+        <mat-icon>launch</mat-icon>
```
Comment (philgetz) [inline src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html:20]: This launch icon should be flexed to the end of the container

File: src/platform/core/dynamic-menu/dynamic-menu.component.html [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  <mat-icon *ngIf="trigger.icon" [class]="trigger.iconClasses">{{ trigger.icon }}</mat-icon>
+</button>
+<button
+  *ngIf="trigger.text"
+  mat-button
+  id="{{ trigger.id }}"
+  [matMenuTriggerFor]="menu.childMenu"
+  [matTooltip]="trigger.tooltip"
+  matTooltipShowDelay="0"
+  matTooltipPosition="below"
```
Comment (philgetz) [inline src/platform/core/dynamic-menu/dynamic-menu.component.html:20]: No need to have a tooltip here, developers should be able to attach it by using the directive on the host. This way we don't introduce an unnecesary dependency.

File: src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html (hunk 1 above)
Comment (thomaslennan) [reply src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html]: Thanks for helping with the styling on that. Changed. Added margin on left to pad from the text content.

File: src/platform/core/dynamic-menu/dynamic-menu.component.html (hunk 2 above)
Comment (thomaslennan) [reply src/platform/core/dynamic-menu/dynamic-menu.component.html]: Removed.

File: src/app/content/components/component-demos/dynamic-menu/demos/dynamic-menu-demo.module.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    /** Angular Modules */
+    CommonModule,
+    MatIconModule,
+    MatButtonModule,
+    MatIconModule,
+    MatButtonModule,
+    MatListModule,
+    MatTabsModule,
+    MatToolbarModule,
+    /** Hero Module */
```
Comment (christianmemije) [inline src/app/content/components/component-demos/dynamic-menu/demos/dynamic-menu-demo.module.ts:32]: Hero?

File: src/platform/core/dynamic-menu/README.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
+# td-dynamic-menu
+
+`td-dynamic-menu` element supports the creation of n-level deep cascading menus with links using a JSON definition language.
```
Comment (christianmemije) [inline src/platform/core/dynamic-menu/README.md:3]: `using a JSON definition language` -> `object`

File: src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html [hunk 5]
```diff
@@ -0,0 +1,30 @@
+<mat-menu #childMenu="matMenu" [overlapTrigger]="false">
+  <ng-template let-item ngFor [ngForOf]="items">
+    <ng-container *ngIf="item.children && item.children.length > 0">
+      <button mat-menu-item id="{{ item.id }}" [matMenuTriggerFor]="menu.childMenu">
```
Comment (christianmemije) [inline src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html:4]: let's change this to [id]="item.id"

File: src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+          {{ item.text }}
+        </span>
+      </button>
+      <td-dynamic-menu-item #menu [items]="item.children"></td-dynamic-menu-item>
+    </ng-container>
+    <ng-container *ngIf="!item.children || item.children.length === 0">
+      <a
+        class="new-tab"
+        mat-menu-item
+        id="{{ item.id }}"
```
Comment (christianmemije) [inline src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html:17]: let's change this to [id]="item.id"

File: src/app/content/components/component-demos/dynamic-menu/demos/dynamic-menu-demo-basic/dynamic-menu-demo-basic.component.html [hunk 7]
```diff
@@ -0,0 +1,4 @@
+<div layout="column">
```
Comment (christianmemije) [inline src/app/content/components/component-demos/dynamic-menu/demos/dynamic-menu-demo-basic/dynamic-menu-demo-basic.component.html:1]: Let's change this to row? Or remove in that case. The alignment looks a bit off when stacked.

File: src/app/content/components/component-demos/dynamic-menu/demos/dynamic-menu-demo.module.ts (hunk 3 above)
Comment (thomaslennan) [reply src/app/content/components/component-demos/dynamic-menu/demos/dynamic-menu-demo.module.ts:32]: Removing. Copy/paste from another demo.

File: src/platform/core/dynamic-menu/README.md (hunk 4 above)
Comment (thomaslennan) [reply src/platform/core/dynamic-menu/README.md:3]: Switched.

File: src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html (hunk 5 above)
Comment (thomaslennan) [reply src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html:4]: Switched.

File: src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html (hunk 6 above)
Comment (thomaslennan) [reply src/platform/core/dynamic-menu/dynamic-menu-item/dynamic-menu-item.component.html:17]: Switched.

File: src/app/content/components/component-demos/dynamic-menu/demos/dynamic-menu-demo-basic/dynamic-menu-demo-basic.component.html (hunk 7 above)
Comment (thomaslennan) [reply src/app/content/components/component-demos/dynamic-menu/demos/dynamic-menu-demo-basic/dynamic-menu-demo-basic.component.html:1]: OK

**Pull request merged**